nixos/modules/services/web-servers/keter/config.nix

```diff
@@ -0,0 +1,87 @@
+{ globalketerConfig
+, keterRoot
+, writeShellScript
+, writeTextFile
+, lib
+, bundleCfg
+, callPackage
+, coreutils
+, keter
+, ...
+}:
+
+let
+  incoming = "${keterRoot}/incoming";
+
+
+  globalKeterConfigFile = writeTextFile {
+    name = "keter-config.yml";
+    text = (lib.generators.toYAML { } globalketerConfig);
+  };
+
+  # If things are expected to change often, put it in the bundle!
+  bundle = callPackage ./bundle.nix
+    (bundleCfg // { inherit executable; });
+
+  # This indirection is required to ensure the nix path
+  # gets copied over to the target machine in remote deployments.
+  # Furthermore, it's important that we use exec to
+  # run the binary otherwise we get process leakage.
+  executable = writeShellScript "bundle-wrapper" ''
+    #!/usr/bin/env bash
+    export PATH=$PATH:${lib.makeBinPath [ coreutils ]}
+    set -e
+    ${bundleCfg.secretScript}
+    set -xe
+    ${bundleCfg.publicScript}
+    exec ${bundleCfg.executable}
+  '';
+
+in
+{
+  # This is keter, which loads our app on deploy, and checks if it is healthy
+  # be very careful putting anything in here that changes,
+  # any change to the systemd unit can cause brief windows of downtime
+  # because nixos has to restart the unit.
+  keter = {
+    description = "keter app loader";
+    script = ''
+      set -xe
+      mkdir -p ${incoming}
+      { tail -F /opt/keter/log/keter/current.log -n 0 & ${keter}/bin/keter ${globalKeterConfigFile}; }
```

Comment:
Why `/opt`? This is (almost) not a thing in NixOS.

Also, does `tail` die with the rest of the service? Check `journalctl` as it might be complaining about that on service restart.

Reply:
good catch, that `/opt` should be keterRoot instead, it's by default `/opt/keter` because that's how snoyman described it in the readme before I took over as maintainer.
Perhaps there would be a better default for nixos?

--- 

about killing, here I'm `journalctl -fu keter`, the first two times I used systemctl restart, the last time I used pkill keter (to simulate a a crash), looks like it's just functioning as expected, it reboots with no complaints about tail as far as I can tell.

```
May 25 09:43:30 supercede_staging systemd[1]: Stopping keter app loader...
May 25 09:43:30 supercede_staging systemd[1]: keter.service: Deactivated successfully.
May 25 09:43:30 supercede_staging systemd[1]: Stopped keter app loader.
May 25 09:43:30 supercede_staging systemd[1]: keter.service: Consumed 6h 39min 1.650s CPU time, received 795.8M IP traffic, sent 876.6M IP traffic.
May 25 09:43:30 supercede_staging systemd[1]: Started keter app loader.
May 25 09:43:30 supercede_staging keter-start[3922060]: + mkdir -p /opt/keter/incoming
May 25 09:43:30 supercede_staging keter-start[3922060]: + /nix/store/pgrcw6dkpk8d3lwanby07xk6m2gx956c-keter-2.0.1/bin/keter /nix/store/25dmvgdl838a5dlq29jxq1cnkjdgr8ff-keter-config.yml
May 25 09:43:30 supercede_staging keter-start[3922062]: + tail -F /opt/keter/log/keter/current.log -n 0
May 25 09:43:30 supercede_staging keter-start[3922062]: tail: '/opt/keter/log/keter/current.log' has become inaccessible: No such file or directory
May 25 09:43:30 supercede_staging keter-start[3922062]: tail: '/opt/keter/log/keter/current.log' has appeared;  following new file
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Launching cli
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Launching initial
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/myapp.keter\" 1653300542"
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Started watching
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Unpacking bundle '/opt/keter/incoming/myapp.keter'
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/supercede.keter\" 1653410590"
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Started listening
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.19: Unpacking bundle '/opt/keter/incoming/supercede.keter'
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Sanity checks passed
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Sanity checks passed
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Reserving hosts for app myapp: staging.supercede.com
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Reserving hosts for app supercede: staging.supercede.com
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Process created: /nix/store/73hjl15h722fxbdiap547mxzrsy2nalp-bundle-wrapper
May 25 09:43:30 supercede_staging keter-start[3922062]: 2022-05-25 09:43:30.20: Error occured when launching bundle supercede: CannotReserveHosts supercede (fromList [("staging.supercede.com",myapp)])
May 25 09:43:34 supercede_staging keter-start[3922062]: 2022-05-25 09:43:34.20: Activating app myapp with hosts: staging.supercede.com
May 25 09:43:59 supercede_staging systemd[1]: Stopping keter app loader...
May 25 09:43:59 supercede_staging systemd[1]: keter.service: Deactivated successfully.
May 25 09:43:59 supercede_staging systemd[1]: Stopped keter app loader.
May 25 09:43:59 supercede_staging systemd[1]: keter.service: Consumed 16.132s CPU time, received 528.3K IP traffic, sent 673.9K IP traffic.
May 25 09:43:59 supercede_staging systemd[1]: Started keter app loader.
May 25 09:43:59 supercede_staging keter-start[3922122]: + mkdir -p /opt/keter/incoming
May 25 09:43:59 supercede_staging keter-start[3922122]: + /nix/store/pgrcw6dkpk8d3lwanby07xk6m2gx956c-keter-2.0.1/bin/keter /nix/store/25dmvgdl838a5dlq29jxq1cnkjdgr8ff-keter-config.yml
May 25 09:43:59 supercede_staging keter-start[3922124]: + tail -F /opt/keter/log/keter/current.log -n 0
May 25 09:43:59 supercede_staging keter-start[3922124]: tail: '/opt/keter/log/keter/current.log' has become inaccessible: No such file or directory
May 25 09:43:59 supercede_staging keter-start[3922124]: tail: '/opt/keter/log/keter/current.log' has appeared;  following new file
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Launching cli
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Launching initial
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/myapp.keter\" 1653300542"
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Started watching
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Unpacking bundle '/opt/keter/incoming/myapp.keter'
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/supercede.keter\" 1653410590"
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Unpacking bundle '/opt/keter/incoming/supercede.keter'
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.67: Started listening
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Sanity checks passed
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Reserving hosts for app myapp: staging.supercede.com
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Sanity checks passed
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Reserving hosts for app supercede: staging.supercede.com
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Error occured when launching bundle supercede: CannotReserveHosts supercede (fromList [("staging.supercede.com",myapp)])
May 25 09:43:59 supercede_staging keter-start[3922124]: 2022-05-25 09:43:59.68: Process created: /nix/store/73hjl15h722fxbdiap547mxzrsy2nalp-bundle-wrapper
May 25 09:44:03 supercede_staging keter-start[3922124]: 2022-05-25 09:44:03.68: Activating app myapp with hosts: staging.supercede.com
May 25 09:44:54 supercede_staging systemd[1]: keter.service: Deactivated successfully.
May 25 09:44:54 supercede_staging systemd[1]: keter.service: Consumed 21.838s CPU time, received 798.0K IP traffic, sent 985.2K IP traffic.
May 25 09:45:04 supercede_staging systemd[1]: keter.service: Scheduled restart job, restart counter is at 1.
May 25 09:45:04 supercede_staging systemd[1]: Stopped keter app loader.
May 25 09:45:04 supercede_staging systemd[1]: keter.service: Consumed 21.838s CPU time, received 798.0K IP traffic, sent 985.2K IP traffic.
May 25 09:45:04 supercede_staging systemd[1]: Started keter app loader.
May 25 09:45:04 supercede_staging keter-start[3922208]: + mkdir -p /opt/keter/incoming
May 25 09:45:04 supercede_staging keter-start[3922208]: + /nix/store/pgrcw6dkpk8d3lwanby07xk6m2gx956c-keter-2.0.1/bin/keter /nix/store/25dmvgdl838a5dlq29jxq1cnkjdgr8ff-keter-config.yml
May 25 09:45:04 supercede_staging keter-start[3922210]: + tail -F /opt/keter/log/keter/current.log -n 0
May 25 09:45:04 supercede_staging keter-start[3922210]: tail: '/opt/keter/log/keter/current.log' has been replaced;  following new file
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Launching cli
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Launching initial
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/myapp.keter\" 1653300542"
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Unpacking bundle '/opt/keter/incoming/myapp.keter'
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Started watching
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Reloading from: Nothing to "AIBundle \"/opt/keter/incoming/supercede.keter\" 1653410590"
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Sanity checks passed
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Unpacking bundle '/opt/keter/incoming/supercede.keter'
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Reserving hosts for app myapp: staging.supercede.com
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.24: Started listening
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Sanity checks passed
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Process created: /nix/store/73hjl15h722fxbdiap547mxzrsy2nalp-bundle-wrapper
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Reserving hosts for app supercede: staging.supercede.com
May 25 09:45:04 supercede_staging keter-start[3922210]: 2022-05-25 09:45:04.25: Error occured when launching bundle supercede: CannotReserveHosts supercede (fromList [("staging.supercede.com",myapp)])
May 25 09:45:08 supercede_staging keter-start[3922210]: 2022-05-25 09:45:08.25: Activating app myapp with hosts: staging.supercede.com
```

Reply:
`/var/lib/keter` seems to be the consensus, on my current machine:
```
ckie@cookiemonster ~ -> ls /var/lib
AccountsService   alsa  cntr        cups      dnscrypt-proxy  lightdm       logrotate.status  misc            NetworkManager-fortisslvpn  private  tailscale
adguard-dnsproxy  cni   containers  dhclient  libvirt         lightdm-data  machines          NetworkManager  nixos                       systemd  udisks2
ckie@cookiemonster ~ -> ls /opt
"/opt": No such file or directory (os error 2)
```